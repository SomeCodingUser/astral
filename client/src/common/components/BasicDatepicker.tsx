import { FC, useEffect, useState } from 'react'
import Datepicker from 'tailwind-datepicker-react'
import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import { CalendarIcon } from '@heroicons/react/24/outline'
import { IOptions } from 'tailwind-datepicker-react/types/Options'

dayjs.extend(utc)

type Props = {
  value?: string | undefined
  onChange?: (selectedDate: Date) => void
  minDate?: Date
}

const defaultOptions = {
  theme: {
    background: 'bg-white dark:bg-[#1E254E] dark:text-white',
    disabledText: 'text-gray-500/75 dark:text-gray-400/75',
    input: 'py-3 w-full rounded-[20px] bg-white dark:bg-[#1E254E] dark:text-white',
    todayBtn: '',
    clearBtn: '',
    icons: '',
    text: '',
    inputIcon: '',
    selected: '',
  },
}

const BasicDatepicker: FC<Props> = ({ onChange, value, minDate }) => {
  const [show, setShow] = useState<boolean>(false)
  const [options, setOptions] = useState<IOptions>(defaultOptions)

  const handleClose = (state: boolean) => {
    setShow(state)
  }

  useEffect(() => {
    if (minDate) {
      setOptions({ ...defaultOptions, minDate: new Date(minDate), maxDate: new Date(minDate) })
    }
  }, [minDate])

  return (
    <Datepicker options={options} onChange={onChange} show={show} setShow={handleClose}>
      <div className='relative'>
        <span className='absolute flex py-[22px] left-3 items-center h-5 w-5 text-[#282929] dark:text-white'>
          <CalendarIcon />
        </span>
        <input
          type='text'
          placeholder='Select Date'
          value={value ? dayjs(value).utc().format('MMMM DD, YYYY') : ''}
          onFocus={() => setShow(!show)}
          className='px-10 py-3 w-full rounded-[42px] bg-white dark:bg-[#1E254E] dark:text-white border-transparent focus:border-gray-500 focus:bg-white focus:ring-0 text-sm'
          readOnly
        />
      </div>
    </Datepicker>
  )
}

export default BasicDatepicker
