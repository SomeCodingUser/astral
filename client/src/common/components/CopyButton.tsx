import { ButtonHTMLAttributes, FC, ReactNode } from 'react'
import { Square2StackIcon } from '@heroicons/react/24/outline'
import toast from 'react-hot-toast'

// common
import { useClipboard } from 'common/hooks/useClipboard'

interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
  value: string
  message?: string
  children?: ReactNode
  buttonClass?: string
  iconClass?: string
}

const CopyButton: FC<Props> = ({
  value,
  children,
  message = 'Copied',
  buttonClass = 'w-full',
  iconClass = 'w-4 h-4',
  ...rest
}) => {
  const { onCopy } = useClipboard(value)

  const handleCopyClick = (): void => {
    onCopy()
    toast.success(message, {
      position: 'bottom-center',
    })
  }

  return (
    <button className={`flex gap-2 ${buttonClass}`} onClick={handleCopyClick} {...rest}>
      {children}
      <Square2StackIcon className={iconClass} stroke='#DE67E4' />
    </button>
  )
}

export default CopyButton
