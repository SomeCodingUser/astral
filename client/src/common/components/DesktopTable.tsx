import { ChevronDownIcon, ChevronUpDownIcon, ChevronUpIcon } from '@heroicons/react/20/solid'
import { Table, flexRender } from '@tanstack/react-table'

interface TableProps<T extends object> {
  table: Table<T>
}

const DesktopTable = <T extends object>({ table }: TableProps<T>) => (
  <table className="min-w-max table-auto font-['Montserrat'] bg-white rounded-[20px] dark:bg-gradient-to-r dark:from-[#4141B3] dark:via-[#6B5ACF] dark:to-[#896BD2] dark:border-none">
    <thead className='text-[#857EC2] text-sm dark:text-white/75 border-b border-gray-200'>
      {table.getHeaderGroups().map((headerGroup) => (
        <tr key={headerGroup.id}>
          {headerGroup.headers.map((header) => (
            <th
              key={header.id}
              {...(header.column.getCanSort()
                ? { onClick: header.column.getToggleSortingHandler() }
                : {})}
              className='px-3 py-4 text-sm font-normal text-start'
            >
              <div className='flex gap-1 justify-items-center align-middle'>
                {header.isPlaceholder
                  ? null
                  : flexRender(header.column.columnDef.header, header.getContext())}
                {header.column.getCanSort() ? (
                  header.column.getIsSorted() === 'asc' ? (
                    <span className='cursor-pointer relative inline-flex items-center rounded-full text-sm font-medium text-[#DE67E4]  focus:z-20 mr-[14px] dark:text-white  dark:border-none'>
                      <span className='sr-only'>Up</span>
                      <ChevronUpIcon className='h-5 w-5' aria-hidden='true' />
                    </span>
                  ) : header.column.getIsSorted() === 'desc' ? (
                    <span className='cursor-pointer relative inline-flex items-center rounded-full text-sm font-medium text-[#DE67E4] focus:z-20 mr-[14px] dark:text-white  dark:border-none'>
                      <span className='sr-only'>Down</span>
                      <ChevronDownIcon className='h-5 w-5' aria-hidden='true' />
                    </span>
                  ) : (
                    <span className='cursor-pointer relative inline-flex items-center rounded-full text-sm font-medium text-[#DE67E4]  focus:z-20 mr-[14px] dark:text-white  dark:border-none'>
                      <span className='sr-only'>UpDown</span>
                      <ChevronUpDownIcon className='h-5 w-5' aria-hidden='true' />
                    </span>
                  )
                ) : (
                  <span></span>
                )}
              </div>
            </th>
          ))}
        </tr>
      ))}
    </thead>
    <tbody className='text-gray-600 text-sm font-light dark:text-white'>
      {table.getRowModel().rows.length ? (
        table.getRowModel().rows.map((row, index) => (
          <tr
            key={row.id}
            className={`hover:bg-gray-100 dark:hover:bg-transparent/10 ${
              index !== table.getRowModel().rows.length - 1 && 'border-y border-gray-200'
            } `}
          >
            {row.getVisibleCells().map((cell) => (
              <td className='whitespace-nowrap px-6 py-4 text-sm font-light' key={cell.id}>
                {flexRender(cell.column.columnDef.cell, cell.getContext())}
              </td>
            ))}
          </tr>
        ))
      ) : (
        <div className='w-full flex p-6 hover:bg-gray-100 dark:hover:bg-transparent/10'>
          <div className='text-[#857EC2] text-sm dark:text-white/75'>No entries to show</div>
        </div>
      )}
    </tbody>
  </table>
)

export default DesktopTable
