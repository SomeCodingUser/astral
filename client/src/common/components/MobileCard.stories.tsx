import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import MobileCard from './MobileCard'

export default {
  title: 'Common/MobileCard',
  component: MobileCard,
} as ComponentMeta<typeof MobileCard>

const Template: ComponentStory<typeof MobileCard> = (args) => <MobileCard {...args} />

export const Primary = Template.bind({})
Primary.args = {
  header: 'Header',
  body: [{ name: 'sample name', value: 'sample value' }],
  id: 'sample id',
}
