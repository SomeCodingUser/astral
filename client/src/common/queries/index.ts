import { gql } from '@apollo/client'

export const QUERY_BLOCK_AND_EXTRINSIC_BY_HASH = gql`
  query BlocksAndExtrinsicByHash($hash: String!) {
    blocks(limit: 10, where: { hash_eq: $hash }) {
      id
      height
    }
    extrinsics(limit: 10, where: { hash_eq: $hash }) {
      id
    }
  }
`
