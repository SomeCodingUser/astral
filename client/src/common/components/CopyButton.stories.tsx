import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import CopyButton from './CopyButton'

export default {
  title: 'Common/CopyButton',
  component: CopyButton,
} as ComponentMeta<typeof CopyButton>

const Template: ComponentStory<typeof CopyButton> = (args) => <CopyButton {...args} />

export const Primary = Template.bind({})
Primary.args = {
  value: 'Value to copy',
}
