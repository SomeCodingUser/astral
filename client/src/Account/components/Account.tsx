import { FC } from 'react'
import { useParams } from 'react-router-dom'
import { useQuery } from '@apollo/client'
import { useErrorHandler } from 'react-error-boundary'

// account
import { QUERY_ACCOUNT_BY_ID, QUERY_LATEST_REWARDS } from 'Account/query'
import {
  AccountDetailsCard,
  AccountExtrinsicList,
  AccountGraphs,
  AccountLatestRewards,
} from 'Account/components'

// layout
import { NotFound } from 'layout/components'

// common
import { Spinner } from 'common/components'
import useMediaQuery from 'common/hooks/useMediaQuery'
import { accountIdToHex, formatAddress } from 'common/helpers/formatAddress'

const Account: FC = () => {
  const { accountId } = useParams<{ accountId?: string }>()

  const convertedAddress = formatAddress(accountId)
  const hexAddress = accountIdToHex(accountId || '')

  const isDesktop = useMediaQuery('(min-width: 1024px)')

  const { data, error, loading } = useQuery(QUERY_ACCOUNT_BY_ID, {
    variables: { accountId: convertedAddress },
  })

  const {
    data: dataRewards,
    error: errorRewards,
    loading: loadingRewards,
  } = useQuery(QUERY_LATEST_REWARDS, {
    variables: { accountId: hexAddress },
  })

  useErrorHandler(error)
  useErrorHandler(errorRewards)

  if (!convertedAddress) {
    return <NotFound />
  }

  if (loading || loadingRewards) {
    return <Spinner />
  }

  if (!data.accountById || !dataRewards.rewardEvents) {
    return <NotFound />
  }

  const account = data.accountById

  return (
    <div className='w-full flex flex-col space-y-4'>
      <AccountDetailsCard
        account={account}
        accountAddress={convertedAddress}
        isDesktop={isDesktop}
      />
      <div className='flex flex-col lg:flex-row lg:justify-between gap-8'>
        <AccountGraphs hexAddress={hexAddress} account={account} isDesktop={isDesktop} />
        <AccountLatestRewards rewards={dataRewards.rewardEvents} isDesktop={isDesktop} />
      </div>
      <AccountExtrinsicList accountId={convertedAddress} />
    </div>
  )
}

export default Account
