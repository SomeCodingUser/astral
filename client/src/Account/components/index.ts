import AccountList from 'Account/components/AccountList'
import Account from 'Account/components/Account'
import AccountDetailsCard from 'Account/components/AccountDetailsCard'
import AccountDetailsTabs from 'Account/components/AccountDetailsTabs'
import AccountBalancePieChart from 'Account/components/AccountBalancePieChart'
import AccountBalanceStats from 'Account/components/AccountBalanceStats'
import AccountExtrinsicList from 'Account/components/AccountExtrinsicList'
import AccountTable from 'Account/components/AccountTable'
import AccountListCard from 'Account/components/AccountListCard'
import AccountLatestRewards from 'Account/components/AccountLatestRewards'
import AccountGraphs from 'Account/components/AccountGraphs'
import AccountGraphTabs from 'Account/components/AccountGraphTabs'
import AccountRewardGraph from 'Account/components/AccountRewardGraph'
import AccountRewardList from 'Account/components/AccountRewardList'
import AccountRewardListCard from 'Account/components/AccountRewardListCard'
import AccountRewardTable from 'Account/components/AccountRewardTable'

export {
  Account,
  AccountList,
  AccountDetailsCard,
  AccountDetailsTabs,
  AccountBalancePieChart,
  AccountBalanceStats,
  AccountExtrinsicList,
  AccountTable,
  AccountListCard,
  AccountLatestRewards,
  AccountGraphs,
  AccountGraphTabs,
  AccountRewardGraph,
  AccountRewardList,
  AccountRewardListCard,
  AccountRewardTable,
}
