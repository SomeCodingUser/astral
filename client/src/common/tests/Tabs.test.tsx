import { render } from '@testing-library/react'

// common - Tabs
import { Tabs, Tab } from 'common/components'

describe('Tabs', () => {
  it('renders properly', async () => {
    const { asFragment } = render(
      <Tabs>
        <Tab title='Tab title'>Sample content</Tab>
      </Tabs>,
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
