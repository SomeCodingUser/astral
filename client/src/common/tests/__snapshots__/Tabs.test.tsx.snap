// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders properly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap bg-white border border-slate-100 shadow rounded-lg p-4"
  >
    <div
      class="w-full"
    >
      <ul
        class="flex w-full list-none flex-wrap pt-3 pb-4 flex-row "
        role="tablist"
      >
        <li
          class="-mb-px last:mr-0 text-center w-full lg:w-auto"
        >
          <button
            class="w-full lg:w-auto text-xs font-semibold px-[13.8px] py-3 rounded-full block leading-normal text-white bg-[#241235] dark:bg-[#1E254E]"
          >
            Tab title
          </button>
        </li>
      </ul>
      <div>
        Sample content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
